command/agent/command.go

```diff
@@ -978,66 +978,31 @@ func (c *Command) handleRetryJoin(config *Config) error {
 	return nil
 }
 
-// handleSignals blocks until we get an exit-causing signal
-func (c *Command) handleSignals() int {
-	signalCh := make(chan os.Signal, 4)
-	signal.Notify(signalCh, os.Interrupt, syscall.SIGTERM, syscall.SIGHUP, syscall.SIGPIPE)
-
-	// Signal readiness only once signal handlers are setup
-	sdSock, err := openNotify()
-	if err != nil {
-		c.agent.logger.Debug("notify socket could not be accessed", "error", err)
-	}
-	if sdSock != nil {
-		defer sdSock.Close()
-	}
-	sdNotify(sdSock, sdReady)
-
-	// Wait for a signal
-WAIT:
-	var sig os.Signal
-	select {
-	case s := <-signalCh:
-		sig = s
-	case <-winsvc.ShutdownChannel():
-		sig = os.Interrupt
-	case <-c.ShutdownCh:
-		sig = os.Interrupt
-	case <-c.retryJoinErrCh:
-		return 1
-	}
+// terminateGracefully attempts a graceful leave
+func (c *Command) terminateGracefully(signalCh chan os.Signal, sdSock io.Writer) int {
+	sdNotify(sdSock, sdStopping)
 
-	// Skip any SIGPIPE signal and don't try to log it (See issues #1798, #3554)
-	if sig == syscall.SIGPIPE {
-		goto WAIT
-	}
+	gracefulCh := make(chan struct{})
+	defer close(gracefulCh)
 
-	c.Ui.Output(fmt.Sprintf("Caught signal: %v", sig))
+	timeout := gracefulTimeout
 
-	// Check if this is a SIGHUP
-	if sig == syscall.SIGHUP {
-		sdNotifyReloading(sdSock)
-		c.handleReload()
-		sdNotify(sdSock, sdReady)
-		goto WAIT
+	config := c.agent.config
+	if config == nil {
+		c.Ui.Output("Unable to read the agent configuration, using the default graceful timeout")
 	}
 
-	// Check if we should do a graceful leave
-	graceful := false
-	if sig == os.Interrupt && c.agent.GetConfig().LeaveOnInt {
-		graceful = true
-	} else if sig == syscall.SIGTERM && c.agent.GetConfig().LeaveOnTerm {
-		graceful = true
-	}
+	if config.Client != nil && config.Client.Drain != nil && config.Client.Drain.Deadline != nil {
+		ddl, err := time.ParseDuration(*config.Client.Drain.Deadline)
+		if err != nil {
+			c.Ui.Error(fmt.Sprintf("Unable to read the client's drain deadline, using the default graceful timeout: %s", err))
+		}
```

Comment:
Another option would be to look at `c.agent.client.GetConfig().Drain.Deadline` to get the client's already-parsed configuration so we don't have to re-parse the HCL value and we can't ever hit an error.

Reply:
Thanks!! I didn't think about getting it from the client directly instead of the agent